src/Npgsql/Internal/TypeInfoMapping.cs

```diff
@@ -652,18 +653,33 @@ void AddPolymorphicResolverArrayType(TypeInfoMapping elementMapping, Type type,
     }
 
     /// Returns whether type matches any of the types we register pg arrays as.
+    [UnconditionalSuppressMessage("Trimming", "IL2070",
+        Justification = "We care about IList<T> implementations if the instantiation is actually rooted by us through a registered mapping.")]
     public static bool IsArrayLikeType(Type type, [NotNullWhen(true)] out Type? elementType)
     {
-        elementType = type switch
+        if (type.GetElementType() is { } t)
         {
-            { IsArray: true } => type.GetElementType(),
-            { IsConstructedGenericType: true } when type.GetGenericTypeDefinition() is { } def &&
-                                                    (def == typeof(List<>) || def == typeof(IList<>))
-                => type.GetGenericArguments()[0],
-            _ => null
-        };
+            elementType = t;
+            return true;
+        }
+
+        if (type.IsConstructedGenericType && type.GetGenericTypeDefinition() is var def && (def == typeof(List<>) || def == typeof(IList<>)))
```

Comment:
Is this needed given the IList interface check just below, which would also capture it?

Reply:
It's quite a bit faster if we can do it this way. Additionally IList<T>.GetInterfaces() won't return IList<T>